Copter: Auto mode gets attitude control

rmackay9 opened the pull request:
This PR adds support for a new NAV_ATTITUDE_TIME mission command that maintains a specified roll, pitch and yaw target for a specified time.  While auto is in this attitude control submode it is impervious to EKF failsafes.

This is the corresponding mavlink PR https://github.com/ArduPilot/mavlink/pull/272

This has been tested in SITL consisting of a circuit of WAYPOINT commands and a few NAV_ATTITUDE_TIME commands which caused the vehicle to fly East for 60 seconds and then fly West for 60 seconds before resuming the circuit near home.  During this test the GPS was disabled using the RCx_OPTION = 65 and all worked as expected:

- Disabling the GPS had no effect while the vehicle was executing the NAV_ATTITUDE_TIME command
- If the GPS recovered before the next WAYPOINT command the vehicle continued the circuit
- If the GPS remained disabled when the next WAYPOINT command started the EKF failsafe triggered a LAND

Below are some screenshots of the SITL testing

![image](https://user-images.githubusercontent.com/1498098/170420294-96ebdad0-be2e-4547-a463-8792a1f93827.png)
![image](https://user-images.githubusercontent.com/1498098/170420333-30b550bd-4c02-469e-bf89-3598a060429d.png)
![image](https://user-images.githubusercontent.com/1498098/170420380-f2d9233b-9788-4856-9784-665fe580cb75.png)

How to Use
------------

- Create an Auto mission with the regular Takeoff, RTL, etc
- If using MP, add an UNKNOWN command and set the "command ID" to 42703
    - 1st column is time in seconds
    - 2nd column = roll (in degrees, positive is roll right, negative is roll left)
    - 3rd column = pitch (in degrees, negative is pitch forward, positive is pitch back)
    - 4th column is yaw (in degrees, 0=North, 90=East, 180=South, 270=West, all inbetween values also OK)
    - 5th column is climb rate (in m/s, positive is up, negative is down) <-- best to leave as zero
- arm the vehicle and switch to Auto
- optionally try disabling the GPS while the vehicle is executing the new command by setting RCx_OPTION = 65 and pull the appropriate switch high
![image](https://user-images.githubusercontent.com/1498098/170484745-70e0a39f-c5e6-4631-9f96-418d393194a7.png)

IamPete1 commented:
What's the use case?

pompecukor commented:
> 

For example being able to fly a recon (ISR) mission in area of probable active GPS spuffing (feeding false data). Where you want to intentionally turn of the GPS and want a "go there and return" type mission. With the understanding that there will be error build up with the wind, but you can still have a degree of sucsess. When the drone is "back" even if there a km offset from the home position, you would still be able to turn the GPS back on (with RC or script) and then get correct pos lock and fly home. Or even use fpv camera.

rmackay9 commented:
@IamPete1,

I was thinking of any case where you know there could be GPS interference in a particular area and the user wants to vehicle to just fly through it.  So for example, under a suspension bridge.  This low level angle control could also be useful if you just want the vehicle to drift with the wind for 10 seconds.  Angle control is really the lowest level control so in some ways it's odd that we haven't had it up until now.

I was actually thinking of other varieties like angle-thrust-time (e.g. stabilize-in-auto)  which would not attempt to maintain altitude but instead just maintain a thrust which could be useful for dropping copters out of planes but I figured I'd leave that until someone specifically asks for it.

rmackay9 commented:
I've update this PR in 3 ways:

1. rebased on master
2. updated the mavlink reference so it points to master (and this PR is thus merge-able)
3. added angle limits to ensure the vehicle can maintain the specified climb rate

lthall commented on ArduCopter/mode_auto.cpp:
```diff
@@ -1084,6 +1130,35 @@ void ModeAuto::loiter_to_alt_run()
+    // limit and scale lean angles
+    const float angle_limit_cd = MAX(1000.0f, MIN(copter.aparm.angle_max, attitude_control->get_althold_lean_angle_max_cd()));
+    Vector2f target_rp_cd(nav_attitude_time.roll_deg * 100, nav_attitude_time.pitch_deg * 100);
+    target_rp_cd.limit_length(angle_limit_cd);
```
This will work but could be done better. The ideal implementation would limit the lean angle in the direction of lean without changing the direction. I can fix this in a follow up PR.

pompecukor commented:
Tested on real copter. Seems to work great. Opens a lot of possibilities. @rmackay9 has the logs. Great job. Thank you

rmackay9 commented:
Merged, thanks for the comments and review!
